File: src/investments/non-registered-savings-plan.ts
// tslint:disable:max-line-length
/*
Sources
   TBD

Revised
    2025-12-21
*/
// tslint:enable:max-line-length

import { Rate } from '../taxes';

export const CAPITAL_GAINS_BRACKETS: Rate[] = [
    {
        FROM: 0,
        TO: Infinity,
        RATE: 0.5,
    },
];

export function getCapitalGainsTaxableAmount(totalCapitalGains: number): number {
    return CAPITAL_GAINS_BRACKETS.reduce((taxableAmount, currentBracket) => {
        const currentBracketRate = currentBracket.RATE;

        const bracketTaxableAmount = Math.min(totalCapitalGains, currentBracket.TO);

        if (bracketTaxableAmount <= currentBracket.FROM) return taxableAmount;

        return taxableAmount + ((bracketTaxableAmount - currentBracket.FROM) * currentBracketRate);
    }, 0);
}

File: src/commonMain/kotlin/com/equisoft/taxca/investments/NonRegisteredSavingsPlan.kt
/*
Sources
   TBD

Revised
    2025-12-21
*/

package com.equisoft.taxca.investments

import com.equisoft.taxca.taxes.Rate
import kotlin.math.min

val CapitalGainsBrackets: List<Rate> = listOf(
    Rate(
        from = 0.0,
        to = Double.POSITIVE_INFINITY,
        rate = 0.5,
    ),
)

fun getCapitalGainsTaxableAmount(totalCapitalGains: Double): Double =
    CapitalGainsBrackets.fold(0.0) { taxableAmount, currentBracket ->
        val currentBracketRate = currentBracket.rate

        val bracketTaxableAmount = min(totalCapitalGains, currentBracket.to)

        if (bracketTaxableAmount <= currentBracket.from) {
            taxableAmount
        } else {
            taxableAmount + ((bracketTaxableAmount - currentBracket.from) * currentBracketRate)
        }
    }
